include/nbl/system/IAsyncQueueDispatcher.h

```diff
@@ -15,63 +15,79 @@ namespace impl
     public:
         IAsyncQueueDispatcherBase() = default;
         ~IAsyncQueueDispatcherBase() = default;
+        // dont want to play around with relaxed memory ordering yet
         struct request_base_t
         {
+                enum E_STATE : uint32_t
+                {
+                    ES_INITIAL=0,
+                    ES_RECORDING=1,
+                    ES_PENDING=2,
+                    ES_EXECUTING=3,
+                    ES_READY=4
+                };
+                request_base_t() : state(ES_INITIAL)
+                {
+                    assert(std::atomic_uint32_t::is_lock_free());
+                }
                 ~request_base_t() = default;
 
                 // lock when overwriting the request
-                std::unique_lock<std::mutex> reset()
+                void reset()
                 {
-                    std::unique_lock<std::mutex> lk(mtx);
-                    ready.store(false);
-                    return lk;
+                    transition(ES_INITIAL,ES_RECORDING);
                 }
                 // unlock request after we've written everything into it
-                void finalize(std::unique_lock<std::mutex>&& lk)
+                void finalize()
                 {
-                    lk.unlock();
-                    ready_for_work.store(true);
-                    #if __cplusplus >= 202002L
-                        ready_for_work.notify_one(); // does an atomic need to be notified under a lock?
-                    #endif
+                    const auto prev = state.exchange(ES_PENDING);
+                    assert(prev==ES_RECORDING);
+                    state.notify_one();
                 }
                 // do NOT allow canceling of request while they are processed
-                std::unique_lock<std::mutex> wait_for_work()
+                void wait_for_work()
```

Comment:
isnt this name misleading? i mean it waits for PENDING and then changes to EXECUTING, i'd expect just waiting for some state

Reply:
I guess, but right now the focus isnt on how well semantically named my methods are... rather its whether my solution will work at all.